---
title: "Analysis on pairwise competition assays"
author: "Chang-Yu Chang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
    number_sections: no
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = FALSE)
library(tidyverse)
library(data.table)
library(cowplot)
```


# Goal {-} 

- Combine and read data csv from Rmd 02A to 02D

- Match and compare isolates' information to outcome of pairwise competition
    - RDP taxonomy (e.g., family and genus)
    - 16S differences
    - Distance on phylogenetic tree
    - Monoculture growth rate 
  

# Read and combine pairs data 

```{r}
source("script/02-pairs-01-read_data.R")
```

186 pairs

```{r}
pairs
```


Pairs dataframe melted by T0/T8 and three initial frequencies

```{r}
pairs_melted
```

# Plot outcomes types

```{r warning = F}
source("script/02-pairs-02-outcome_types.R")
```

Four example outcomes: bistability, coexistence, exclusion, and neutrality 

```{r}
# Plot pairs example dynamics
p_pairs_example_outcomes <- pairs_example_outcomes %>%
  plot_pairs_freq(pairs_freq_df = pairs_freq, show_strip = F) +
  facet_grid(.~InteractionType)

# The frequencies of three outcomes
p_pairs_interaction <- plot_bar(pairs, bar_by = "InteractionType") +
  geom_text(data = data.frame(InteractionType = "neutrality", x = 1, y = 1, label = "n = 186 pairs"), mapping = aes(x = x, y = y, label = label))
```

```{r fig.width = 4, fig.height = 3}
p <- plot_grid(p_pairs_interaction, p_pairs_example_outcomes, ncol = 1, align = "v", rel_heights = c(1, 0.5))
p
ggsave(here::here("output/report/figure/02-pairs-pairs_outcomes.pdf"), p, width = 8, height = 6)
```


Four example outcomes: bistability, coexistence, exclusion, and neutrality 

```{r warning = F}
# Plot pairs example dynamics
p_pairs_example_outcomes_finer <- pairs_example_outcomes_finer %>%
  plot_pairs_freq(pairs_freq_df = pairs_freq, show_strip = F) +
  facet_grid(.~InteractionTypeFiner)

# The frequencies of five finer outcomes
p_pairs_interaction_finer <- plot_bar(pairs, bar_by = "InteractionTypeFiner", fill_by = "InteractionType", show_fill_legend = T)

```

```{r fig.width = 5, fig.height = 3}
# Combine plot
#theme_set(theme_cowplot(font_size=12))
p <- plot_grid(p_pairs_interaction_finer, p_pairs_example_outcomes_finer, ncol = 1, align = "v", rel_heights = c(1, 0.4))
p
ggsave(here::here("output/report/figure/02-pairs-pairs_outcomes.pdf"), p, width = 8, height = 6)
```


# Pairs taxonomy vs. outcome of pairwise competition

```{r}
source("script/02-pairs-03-taxonomy.R")
```


## RDP taxonomy

Fermenter vs. Interaction type

```{r fig.width=3, fig.height=3}
p_pairs_interaction_fermenter
ggsave(here::here("output/report/figure/02-pairs-fermenter_interaction.pdf"), p_pairs_interaction_fermenter, width = 5, height = 5)

```

Family vs. Interaction type

```{r fig.width=3, fig.height=3}
p_pairs_interaction_family
ggsave(here::here("output/report/figure/02-pairs-family_interaction.pdf"), p_pairs_interaction_family, width = 5, height = 5)
```

## Pairwise 16S differences

Pairwise 16S differences

```{r}
pairs %>%
  filter(!is.na(SeqDifference)) %>% 
  ggplot(aes(x = SeqDifference, fill = PairFermenter)) +
  geom_histogram(col = "black", binwidth = 2) +
  theme_bw() +
  theme(legend.position = "top") +
  labs(x = "pairwise 16S base pair differences") +
  ggtitle("")
```

Coarsed-grained 16S sequence difference

```{r}
pairs %>%
  filter(!is.na(SeqDifference)) %>% 
  ggplot(aes(x = SeqDifference, fill = PairFermenter)) +
  geom_histogram(col = "black", binwidth = 10) +
  theme_bw() +
  theme(legend.position = "top") +
  labs(x = "pairwise 16S base pair differences") +
  ggtitle("")

```


```{r include = F}
# 16S sequence difference vs. Interaction type
pairs_interaction_16S_cut %>%
  ggplot(aes(x = SeqDifference, y = FracCoext)) +
  geom_col(col = "black", fill = "white") +
  theme_bw() +
  theme(legend.position = "top")
```


```{r fig.width=5, fig.height=5}
p_pairs_interaction_16S_cut
```



## Pairwise distance difference of tips on a tree (deprecated)

Pairwise tree distance difference

```{r}
pairs %>%
  filter(!is.na(PairTreeDistance)) %>% 
  ggplot(aes(x = PairTreeDistance, fill = PairFermenter)) +
  geom_histogram(col = "black") +
  theme_bw() +
  theme(legend.position = "top") +
  labs(x = "pairwise 16S base pair differences") +
  ggtitle("")
```

Coarse-grained tree distance difference

```{r}
pairs %>%
  filter(!is.na(PairTreeDistance)) %>% 
  ggplot(aes(x = PairTreeDistance, fill = PairFermenter)) +
  geom_histogram(col = "black", binwidth = 0.05) +
  theme_bw() +
  theme(legend.position = "top") +
  labs(x = "pairwise 16S base pair differences") +
  ggtitle("")

```

```{r  fig.width=5, fig.height=5}
p_pairs_interaction_tree_cut
```


# Relative abundance within pairs 

```{r}
source("script/02-pairs-04-pairs_relative_abundance.R")
```


Relative abundance of fermenter vs. non-fermenter pairs

```{r}
pairs_freq_taxonomy_fermenter %>%
  filter(Time == "T8") %>%
  filter(!Isolate1MeasuredFreq %in% c(0,1)) %>%
  ggplot() +
  geom_histogram(aes(x = Isolate1MeasuredFreq), col = "black", fill = "white") +
  facet_grid(PairFermenter~., scale = "free") +
  theme_bw() +
  labs("frequencies of isolate 1 at T8")
```

Relative abundance of Enterobacteriaceae and Pseudomonadaceae pairs

```{r}
pairs_freq_taxonomy_family %>%
  filter(Time == "T8") %>%
  filter(!Isolate1MeasuredFreq %in% c(0,1)) %>%
  ggplot() +
  geom_histogram(aes(x = Isolate1MeasuredFreq), col = "black", fill = "white") +
  facet_grid(PairFamily~., scale = "free") +
  theme_bw() +
  labs("frequencies of isolate 1 at T8")

```



# Fraction of pairwise coexistence within community  

```{r}
communities_pair_coext <- pairs %>%
  mutate(Community = ordered(Community, levels = communities_name)) %>%
  filter(InteractionType != "") %>%
  filter(InteractionType %in% c("exclusion", "coexistence")) %>%
  group_by(Community, InteractionType) %>%
  summarize(Count = n()) %>%
  spread(InteractionType, Count) %>%
  mutate(FracCoext = coexistence / (coexistence + exclusion)) %>%
  ungroup() %>%
  mutate(CommunitySize = community_size)

communities_pair_coext$FracCoext[is.na(communities_pair_coext$FracCoext)] <- 0
```

Fraction of coexistence in 13 communities

```{r}
p_communities_pair_coext <- 
communities_pair_coext %>%
  ggplot() +
  geom_bar(aes(x = Community, y = FracCoext), stat = "identity", col = 1) +
  scale_y_continuous(limits = c(0, 1)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "community", y = "fraction of coexistence")

p_communities_pair_coext
pdf(root$find_file("output/figure/communities_fraction_coexistence.pdf"), width = 5, height = 3); p_communities_pair_coext; invisible(dev.off())
```

# Community hierarchy in communities


From Higgins2017, the competitive score $s_i$ of each strain $i$ was defined as its mean fraction $f_{ij}$ after co-culture with each of the $n-1$ competitor strains:

$$s_i=(\sum_{i\neq j}f_{ij})/(n-1)$$

The hierarchy score $h$ for an n-member netowkr is calculated as:

$$h = \sum_{s_i>s_j}f_{ij}$$

```{r}
communities_hierarchy <- fread(root$find_file("data/temp/communities_hierarchy.csv"))

p_communities_hierarchy <- communities_hierarchy %>%
  ggplot() +
  geom_bar(aes(x = Community, y = HierarchyScore), stat = "identity", col = 1) +
  scale_y_continuous(limits = c(0, 1)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "community", y = "hierarchy score")

p_communities_hierarchy
pdf(root$find_file("output/figure/communities_hierarchy.pdf"), width = 5, height = 3); p_communities_hierarchy; invisible(dev.off())

    
```




# Growth traits vs coexistence

```{r}
source("script/02-pairs-05-growth_rates.R")
```

```{r}
pairs_growth_rate
```

Save pairwise growth traits data 

```{r}
if (write_all_csv) fwrite(pairs_growth_rate, file = root$find_file("data/output/pairs_growth_rate.csv"))
```



```{r fig.width=5, fig.height=10, warning = F}
p_pairs_growth_rate_hist
```


```{r fig.width=5, fig.height=10, warning=F}
p_pairs_growth_rate_density
```

```{r}
pdf(root$find_file("output/figure/pairs_growth_rates_hist.pdf"), width = 10, height = 5); for (i in 1:11) print(p_pairs_growth_rate_hist_list[[i]]); invisible(dev.off())

pdf(root$find_file("output/figure/pairs_growth_rates_density.pdf"), width = 10, height = 5); for (i in 1:11) print(p_pairs_growth_rate_density_list[[i]]); invisible(dev.off())
```







