---
title: "Pairwise phylogenetic difference"
author: "Chang-Yu Chang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
    number_sections: no
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = FALSE)
library(tidyverse)
library(data.table)
library(CASEU)
library(cowplot)
```


# Goal

- Correlate competition result with phylogenetic distance

- Measure the pairwise phylogentic distances by difference in 16S base pairs `pairs_taxonomy`
    - Coarse-grained taxonomy (Family and Gennus)
    - Compute the pairwise sequence differences using `Biostring::pairwiseAlignment()` 
    - Compute pairwise tree distance using `ape::cophenetic.phylo()` on built tree. Try out different tree building methods


# Isolates' RDP taxonomy

```{r warning = F}
source("script/02E-competition_phylogeny-01-pairs_taxonomy.R")
```

```{r}
pairs_taxonomy
```



# Isolates' 16S sequence difference


```{r eval = F}
source("script/02E-competition_phylogeny-02-pairs_16S.R")
```

```{r}
pairs_16S
```

Plot the isolate 16S alignments

```{r eval = F}
msaPrettyPrint(aln, file = "isolates_alignment.pdf", output="pdf", 
  paperHeight = 15, paperWidth = 10,
  showNames="left", askForOverwrite=FALSE, verbose=FALSE)
```


# Isolates' branch distances on phylogenetic tree (deprecated)

```{r warning = F, message = F}
source("script/02E-competition_phylogeny-03-pairs_tree_branch.R")
```

```{r}
pairs_tree_distance
```

```{r}
if (write_all_csv) fwrite(pairs_tree_distance, file = root$find_file("data/temp/pairs_tree_distance.csv"))
```












