---
title: "Analysis on isolate Sanger sequences"
author: "Chang-Yu Chang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
    number_sections: no
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = FALSE, echo = FALSE)
library(tidyverse)
library(data.table)
```

# Goal {-}

- Read and align isolate 16S sequences.
    - Read raw `.ab1` file
    - Merge F and R sequences

- Classify isolate taxonomy using RDP classifier.
    - `rRDP`

- Align seqeunces using MUSCLE.
    - `msa`

- Build phylogenetic tree by using FastTree or RAxML.
    - `ggtree`


# Install packages {-}

For `R >= 3.5.0`, use `BiocManager::install()` instead of `biocLite()` to install Bioconductor-maintained packages.

```{r echo = T, eval = F}
install.packages("BiocManager") # Package for managing and installing Bioconductor packages 
BiocManager::install("Biostrings") # DNA string format
BiocManager::install("DECIPHER") # Dependent packages of sangeranalyseR
BiocManager::install("sangeranalyseR")
BiocManager::install("sangerseqR") # Align forward and forward Sanger sequences
BiocManager::install("rRDP") # ribosomal database project
BiocManager::install("rRDPData") # 16S Reference database for RDP
#BiocManager::install("msa") # Multiple sequence alignment
BiocManager::install("ggtree") # Make figure of phylogenetic tree
install.packages("ape") # Analyses of Phylogenetics and Evolution
```

Attach packages 

```{r message = F, echo = T}
library(DECIPHER)
library(sangeranalyseR)
library(sangerseqR)
library(rRDP)
library(rRDPData)
library(ggtree)
#library(msa)
```


# Read isolate 16S sequences

Read and merge ab1 files

This script reads the raw ab1 file from Sanger sequences and merges the forward and backward sequences. It may take ~20 minutes to read all 420 sequences from 210 isolates.

```{r eval = F}
#' This script takes ~10mins to run
#' Do not run this script if `isolates_16S` is already available. 
source("script/01B-isolate_sanger_seq-01-read_align.R")
```

These isolates include Djordje's isolates from simplicity paper, and Jean's isolates from random environments.

# Taxonomy classifier

Use RDP taxononmy classifier to determine the isolates' taxonomy. 

```{r}
# This may take a few seconds
source("script/01B-isolate_sanger_seq-02-RDP.R")
```

```{r}
isolates_RDP
```

# Multiple sequence alignment

Read isolate RDP

```{r}
isolates_RDP <- fread(here::here("data/temp/", "isolates_RDP.csv"))
```

Make DNA string sets

```{r}
isolates_seq <- DNAStringSet(isolates_RDP$Sequence)
names(isolates_seq) <- isolates_RDP$ExpID
isolates_seq
```

Multiple alignment on isolates across all communities. It takes a few seconds.

```{r}
#aln <- msa::msa(isolates_seq)
```

# Build basic tree

## Build tree

```{r}
aln2 <- msaConvert(aln, type = "seqinr::alignment")
d <- seqinr::dist.alignment(aln2, "identity")
tree <- ape::nj(d) # Neighbor-joining tree estimation
```

## Plot basic tree


```{r fig.width=10, fig.height=10}
ggtree(tree) + ggplot2::xlim(0, 0.5)
```

16S phylogeny by fucntion (glucose fermenter)

```{r fig.width=10, fig.height=10, warning=F}
isolates_RDP_p1 <- isolates_RDP %>%
  mutate(TreeTip = paste0(Community, " ", Genus))

p1 <- ggtree(tree) %<+% isolates_RDP_p1 +
  geom_tiplab(aes(label = TreeTip, color = Fermenter), align = T)  +
  ggplot2::xlim(0, 0.6) +
  theme(legend.position = "top") + 
  NULL

p1
if (write_all_pdf == T) pdf(root$find_file("output/report/figure/01B-isolates_sanger_seq-isolate_phylogeny_fermenter.pdf"), height = 15, width = 10); p1; invisible(dev.off())
```

16S phylogeny by community that the isolates are from.

```{r fig.width=10, fig.height=10, warning=F}
isolates_RDP_p2 <- isolates_RDP %>%
  group_by(Community) %>%
  mutate(Isolate = 1:n(), TreeTip = paste0(Community, "\tisolate", Isolate, "\t", Genus))

p2 <- ggtree(tree) %<+% isolates_RDP_p2 +
  geom_tiplab(aes(label = TreeTip, color = Community), align = T)  +
  scale_color_discrete() +
  ggplot2::xlim(0, 0.6) +
  theme(legend.position = "top") + 
  NULL

p2
if (write_all_pdf == TRUE) pdf(root$find_file("output/report/figure/01B-isolates_sanger_seq-isolate_phylogeny_community.pdf"), height = 15, width = 10); p2; invisible(dev.off())
```

67 isolates have Sanger sequences. C2R6 isolate 4 does not have sanger sequence. 68 isolates are used in pairwise competition assays.


## Save msa and tree files

Save the multipple sequence alignment result, ggtree-compatible tree, and the isolate RDP information with the tree in `data/temp/isolates_sanger_seq.Rdata`.

```{r eval = F}
save(aln, aln2, tree, isolates_seq, file = root$find_file("data/temp/isolates_sanger_seq.Rdata"))
```

# Pairwise Sanger sequences difference

```{r}
source("script/01B-isolate_sanger_seq-04-pairwise_alignment.R")
```

Alignemnt all possible pairs in the two isolate sets. The alignment result is later used in `05A-random_network_experiment.Rmd` for assembling the community with unambiguous Sanger sequences 

```{r}
if (write_all_csv) {
  fwrite(pairs_sanger_alignment, root$find_file("data/temp/pairs_sanger_alignment.csv"))
  fwrite(jean_pairs_sanger_alignment, root$find_file("data/temp/jean_pairs_sanger_alignment.csv"))
}
```




















