---
title: ""
author: "Chang-Yu Chang"
date: "`r Sys.Date()`" 
output: 
    pdf_document:
      toc: false
      number_sections: no
header-includes:
    - \usepackage{fancyhdr}
    - \pagestyle{fancy}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = FALSE, echo = FALSE)
library(tidyverse)
library(data.table)
```


# CASEU protocol
`r date = "20191210"`
`r n_plate = 1`

## DNA extraction (Qiagen DNeasy 96)

- Input: pellets from 200 uL T0 mixed inocula, or 500 uL matured culture in DW96s
- Output: 100 uL of DNA in PCR plates

Items                | Usage                             | Amount
---------------------|-----------------------------------|------------
PBS                  | dissolve pellets                  | `r 0.18 * 100 * 2` mL
proteinase K         | DNA extraction                    | `r 0.025 * n_plate * 100` mL
buffer AL            | DNA extraction                    | `r 0.2 * n_plate * 100` mL
ethanol              | DNA extraction                    | `r 0.2 * n_plate * 100` mL
buffer AW1           | DNA extraction                    | `r 0.5 * n_plate * 100` mL
buffer AW2           | DNA extraction                    | `r 0.5 * n_plate * 100` mL
buffer AE            | DNA extraction                    | `r 0.1 * n_plate * 100` mL
DNeasy96 plate       | DNA extraction                    | `r n_plate`  
S-block or DW96      | collect flow-through              | `r n_plate`
PCR plate            | save DNA                          | `r n_plate`
1000uL tips box      | dispense                          | `r n_plate * 3`
200uL tip box        | dispense                          | `r n_plate * 4`
20uL tip box         | dispense                          | `r n_plate * 1`

Table: Materials. \label{material_table}

- Thaw pellet plates at room temperature.
- Dissolve pellets in 180 uL of PBS and transfer to DNeasy microtube plates.
- Add 25 µl proteinase K and 200 µl Buffer AL. Seal the plate with caps and mix by vortexing. Incubate at 56C for 30 min.
- Add 200 µl pure ethanol. Seal the plate with caps and mix thoroughly by vortexing. 
- Assemble the DNeasy 96 plates on top of DW96s. Mark the DNeasy 96 plates.
- Use mP1000 set at 625 uL to transfer the lysate of each sample from DW96s to each well of DNeasy96 plates (maximum 900 uL). 
- Seal DNeasy96 with AirPore film. Use vacuum to collect the flow-through.
- Add 500 uL of Buffer AW1 to each sample. 
- Seal DNeasy96 with AirPore film. Use vacuum to collect the flow-through.
- Add 500 uL of Buffer AW2 to each sample. 
- Seal DNeasy96 with AirPore film. Use vacuum to collect the flow-through.
- Label two PCR plates. Place DNeasy96s on these PCR plates.
- Add 100 ul Buffer AE to each sample, and seal the DNeasy 96 plates with new AirPore Tape Sheets (provided). Incubate for 1 min at room temperature (15–25°C). Centrifuge for 2 min at 6000 rpm.
- Cover the PCR plates with aluminum foil and save it in -20C freezer.


\newpage
## PCR with master mix in 96-well plate

Items                | Usage                             | Amount
---------------------|-----------------------------------|------------
PCR plates           | duplicate PCR                     | `r n_plate * 2`
50 mL faclon tube    | PCR master mix                    | 1
20uL tip box         | dispense                          | `r n_plate * 2 + 1`
200uL tip box        | dispense                          | 1

Table: Materials.

```{r}
# Reagents required per well
data.frame(
  Reagent = c("ddH20", "5X HF buffer", "dNTPs (10mM)", "Phusion"),
  WellVolume = c(23.5, 10, 1, 0.5),
  stringsAsFactors = F
) %>%
  mutate(TotalVolume = WellVolume * n_plate * 108 * 2) %>%
  janitor::adorn_totals() %>% 
  knitr::kable(format = "pandoc", caption = "CASEU PCR master mix.")
```


```{r}
# Reagents used with master mix
data.frame(
  Reagent = c("PCR master mix", "27F primer (3uM)", "1492R primer (3uM)", "teamplate DNA polymerase"),
  WellVolume = c(35, 5, 5, 5), 
  stringsAsFactors = F
)  %>%
  mutate(TotalVolume = WellVolume * n_plate * 108 * 2) %>% # Duplicate PCR
#  mutate(ReservoirWellVolume = TotalVolume / 12) %>%
#  janitor::adorn_totals() %>%  # Add total volume for each well and total reagent used 
  knitr::kable(format = "pandoc", caption = "CASEU PCR reagents with master mix.")
```


- Label `r n_plate * 2` PCR plates with naming convention like **T0 C P2 PCR** and **T3 C P2 PCR**.
- Make **`r total_primer_volume = n_plate * 5 * 108 * 2; total_primer_volume` uL** of 3uM of each primer (`r total_primer_volume * 0.03` uL 100 uM stock + `r total_primer_volume * 0.97` uL ddH2O).
- Premix the PCR reagents (total `r n_plate * 35 * 108 * 2 / 1000` mL) in a 50 mL falcon tube. 
- Use mP200 set at **35 uL** to dispense 35 uL of PCR master mix into each well of `r n_plate * 2` PCR plates. This premix can stay at room temperature.
- Right before starting the PCR reaction, use mP20 set at **5 uL** to add primers.
- Use mP20 set at **5 uL** to add DNA. Cover the PCR plates with clear PCR films. 
- Use the program “CASEU” in “16S” folder. See the table below for PCR cycle.
- Store PCR plates in -20C freezer.


\newpage

```{r}
# Reagents required per well
data.frame(
  Reagent = c("ddH2O", "5X HF buffer", "dNTPs (10mM)", "27F primer (3uM)", "1492R primer (3uM)", "Phusion polymerase", "template DNA"),
  WellVolume = c(23.5, 10, 1, 5, 5, 0.5, 5), 
  stringsAsFactors = F
) %>%
  janitor::adorn_totals() %>% 
  knitr::kable(format = "pandoc", caption = "CASEU PCR reagents.")

```



Step                     |  Temperature | Duration
-------------------------|--------------|-----------
Initial denaturation     | 98 C         | 30 seconds
Amplification (30 cycles)| 98 C         | 30 seconds
\space                   | 50 C         | 30 seconds
\space                   | 72 C         | 90 seconds
Final extension          | 72 C         | 10 minutes
Storage                  | 4 C          | Forever

Table: CASEU PCR cycle. 


\newpage
## SequalPrep PCR product cleanup and normalization

Input: duplicate PCR plates

Output: 20 uL 1.25 ng/uL DNA in PCR plate clean_normalized_PCR


Items                | Usage                             | Amount
---------------------|-----------------------------------|------------
PCR plate            | PCR_mix, unbound_DNA, chean_PCR   | `r n_plate * 3`
SequalPrep plate     | PCR cleanup and normalization     | `r n_plate`
SequalPrep binding buffer | binding                      | `r n_plate * 100 * 0.02` mL
SequalPrep wash buffer    | wash                         | `r n_plate * 100 * 0.05` mL 
SequalPrep elution buffer | elution                      | `r n_plate * 100 * 0.02` mL
20uL tip box         | dispense                          | `r n_plate * 4`
200uL tip box        | dispense                          | `r n_plate * 2`

Table: Materials.

- Thaw the `r n_plate * 2` PCR plates and spin down.
- Label `r n_plate * 3` PCR plates with the naming convention like **T0 C P2 PCR_mix**, **T0 C P2 unbound_DNA**, and **T0 C P2 chean_PCR**. Same for T3.
- Mix 20uL of each replicate PCR into each well of a skirted PCR plate PCR_mix (40 uL in total).
- Add 20uL SequalPrep Binding buffer to each well of SequalPrep plate.
- Add 20uL mixed PCR product into each well of SequalPrep plate.
- Cover with PCR lid, vortex 30 sec, and centrifuge briefly.
- Incubate for 1 hour at room temperature. Note: extra incubation time will not improve yield but will not decrease it either.  Overnight incubations are fine if necessary.
- Transfer unbound DNA to fresh skirted PCR plate unbound_DNA, being careful not to touch the sides of the wells. Discard or foil and save at -20C for up to 30 days and reuse for further SequalPrep.
- Add 50 uL SequalPrep Wash buffer to the SequalPrep plate containing bound DNA and pipet up and down twice to mix. Vortex.
- Remove buffer from wells by pipetting. Tap plate on paper towel gently to remove remaining buffer. Spin down the plate.
- To elute, add 20 uL SequalPrep Elution buffer to each well.
- Cover with lid, vortex 30 sec, and centrifuge briefly.
- Incubate at room temperature for 5 minutes.
- Elute DNA from each well into a new skirted PCR plate clean_normalized_PCR and store at -20C.

\newpage 

## Genewiz Sanger sequencing prep

Input: 20 uL normalized DNA 1.25 ng/uL in clean_normalized_PCR

Output: DNA with primer in a PCR plate for sequencing

Items                | Usage                             | Amount
---------------------|-----------------------------------|------------
PCR plate            | sequencing                        | `r n_plate`
27F 5uM primer       | CASEU primer                      | `r n_plate * 5 * 100 /1000` mL  
20uL tip box         | dispense                          | `r n_plate`

Table: Materials.


- Label `r n_plate` PCR plates with the naming convention **T0 C P2 seq** and **T3 C P2 seq**.
- Make `r n_plate * 5 * 100` uL 5uM 27F primer stock (`r n_plate * 5 * 100 * 0.05` uL 100 uM primer stock + `r n_plate * 5 * 100 * 0.95` uL ddH2O).
- Add **5 uL** of each primer to the PCR plate seq.
- Add **10 uL** of normalized DNA to each well of the PCR plate seq.
- For <48 samples, Label tubes with tube ID from Genewiz online ordering system, which will have your initials and sample number. Write these codes on the sides of tubes. Note: For >48 samples, arrange the samples vertically (in columns).
- Store the PCR seq plates in a box. 
- Drop off the box at Genewiz West Campus pickup location: ISTC room 366.


```{r eval = F}
data.frame(
  DNA = c("1-2ng/uL template DNA", "5uM 27F primer"),
  Volume = c(10, 5),
  FinalMass = c("20 ng", "12 pmol")
) %>%
  mutate(MasterVolume = Volume * 80) %>%
  setNames(c("DNA", "Volume (uL)", "Final Mass", "MasterVolume (uL)")) %>%
  knitr::kable()
```





